resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -211,9 +211,11 @@ private[spark] class ApplicationMaster(
   final def run(): Int = {
     try {
       val attemptID = if (isClusterMode) {
-        // Set the web ui port to be ephemeral for yarn so we don't conflict with
-        // other spark processes running on the same box
-        System.setProperty(UI_PORT.key, "0")
+        // Set the web ui port to be ephemeral for yarn if not set explicitly
+        // so we don't conflict with other spark processes running on the same box
+        if (System.getProperty(UI_PORT.key) != null) {
+          System.setProperty(UI_PORT.key, "0")
+        }
```

Comment:
Hi, @rajatahujaatinmobi . While reviewing this PR, I noticed that SPARK-29465 is a previous one for this issue.

Reply:
yes, please see the discussions there, I think a range of ports would be much better. A single port set on a multi-tenant yarn cluster could easily end up with conflicts. What requirements do you have from Knox, a certain range or ports?
Normally on yarn I would expect you to use the resource manager web ui proxy on a secure cluster so you wouldn't be accessing the UI directly.

Reply:
Correct me if I am wrong

As per the discussions that UI port is always get assigned in the range based on spark.port.maxRetries. So we can always mention the starting port and afterward, it will try to start service based in between (starting port, starting port +spark.port.maxRetries). So we are including the range implicitly. 

Copying the comment
``other port types were spawned in the range (port_mentioned to port_mentioned+spark.port.maxRetries) except the UI port which started on random port.``

We have a Knox proxy that accesses UI via RM link. And Knox proxy has a certain range of ports that it can access on the cluster which is (18000-18159). Unless we start web UI on that range, we can not access the UI. and we can achieve above the range if we set the following properties
spark.ui.port=18000
spark.port.maxRetries=159

Hence setting up a spark.ui.port as starting port is enough instead of setting up a random port number in the range.

Reply:
yea, I think yet Spark could utilize the combination of `spark.ui.port` and `spark.port.maxRetries` to perform "range port".

What else the  "range port" do you refer to? @tgravescs

Reply:
ah yes you are correct you can get range that way. the ports originally wouldn't do the retries or at least we didn't allow configuring maxRetries.

Reply:
I blocked a set of Port range from 18000 to 18050 and configure web or to start at 18008 but it eventually Started at 18051 

**Properties Set: spark.ui.port=18018  spark.port.maxRetries=60**
Successful Case:
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18035. Attempting port 18036.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18036. Attempting port 18037.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18037. Attempting port 18038.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18038. Attempting port 18039.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18039. Attempting port 18040.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18040. Attempting port 18041.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18041. Attempting port 18042.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18042. Attempting port 18043.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18043. Attempting port 18044.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18044. Attempting port 18045.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18045. Attempting port 18046.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18046. Attempting port 18047.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18047. Attempting port 18048.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18048. Attempting port 18049.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18049. Attempting port 18050.
20/06/23 21:55:30 WARN util.Utils: Service 'SparkUI' could not bind on port 18050. Attempting port 18051.
20/06/23 21:55:30 INFO server.AbstractConnector: Started ServerConnector@7ba63fe5{HTTP/1.1,[http/1.1]}{0.0.0.0:18051}
**20/06/23 21:55:30 INFO util.Utils: Successfully started service 'SparkUI' on port 18051.
And  Job succeeded** 

Failure Case 
**Properties Set: spark.ui.port=18018  spark.port.maxRetries=6**

20/06/23 22:16:15 WARN util.Utils: Service 'SparkUI' could not bind on port 18018. Attempting port 18019.
20/06/23 22:16:15 WARN util.Utils: Service 'SparkUI' could not bind on port 18019. Attempting port 18020.
20/06/23 22:16:15 WARN util.Utils: Service 'SparkUI' could not bind on port 18020. Attempting port 18021.
20/06/23 22:16:15 WARN util.Utils: Service 'SparkUI' could not bind on port 18021. Attempting port 18022.
20/06/23 22:16:15 WARN util.Utils: Service 'SparkUI' could not bind on port 18022. Attempting port 18023.
20/06/23 22:16:15 WARN util.Utils: Service 'SparkUI' could not bind on port 18023. Attempting port 18024.
**20/06/23 22:16:15 ERROR ui.SparkUI: Failed to bind SparkUI
java.net.BindException: Address already in use: Service 'SparkUI' failed after 6 retries (starting from 18018)! Consider explicitly setting the appropriate port for the service 'SparkUI' (for example spark.ui.port for SparkUI) to an available port or increasing spark.port.maxRetries.
	at sun.nio.ch.Net.bind0(Native Method)
Job Failed ...**

@tgravescs @Ngone51

Reply:
I don't get your point here. For the successful case, I only see the port range is 18018 ~ 18078 and the port 18051 is valid. Where are those ports 18000, 18050, 18008?

Reply:
I blocked ports from **18000 to 18050 s**o now spark.ui.port can not get port b/w that range. But in Spark properties, I mentioned  **spark.ui.port as 18018 and spark.port.maxRetries=60** so now spark service will start trying from 18018 port until **18018 + 60** in an increment way but ports are blocked until 18050 so spark.ui.port will pick up **18051** port.

Reply:
I see. This is what we expected, right?

Reply:
Yes, that's expected. So I think we are good to go unless someone else has any other doubts.
@Ngone51 @tgravescs

Reply:
thanks for testing that, please update the comment like my below comment requests.